src/codec/encode.jl

```diff
@@ -7,64 +7,52 @@ encode_tag(e::ProtoEncoder, field_number, wire_type::WireType) = encode_tag(e.io
+        # size. If the buffer has `maxsize` set, we use that to estimate the upper bound of the encoded size.
+        # Otherwise, we use 1 byte as that is the mininum size of the encoded size.
+        # If our guess is right, it will save us a copy, but we never want to preallocate too much
+        # space for the size, because then we risk outgrowing the buffer that was allocated with exact size
+        # needed to contain the message.
+        ecoded_size_len_guess = io.maxsize < typemax(Int) ? min(5, _encoded_size(UInt32(io.maxsize - initpos))) : 1
+        truncate(io, initpos + ecoded_size_len_guess)
+        seek(io, initpos + ecoded_size_len_guess)
+        # Now we can encode the object itself
+        f(sink, x, V...) # e.g. _encode(io, x) or _encode(io, x, Val{:zigzag})
         endpos = position(io)
-        data_len = endpos - initpos - MAX_LENGTH_VARINT_BYTES
-        seek(io, initpos)                  # 3.
-        vbyte_encode(io, UInt32(data_len)) # --||--
-        lenght_len = position(io) - initpos
-        unsafe_copyto!(io.data, initpos + lenght_len + 1, io.data, initpos + MAX_LENGTH_VARINT_BYTES + 1, data_len) # 4.
-        seek(io, initpos + lenght_len + data_len)
-        truncate(io, initpos + lenght_len + data_len)
-    else
-        vbyte_encode(io, UInt32(_encoded_size(x)))
-        f(sink, x)
-    end
-    return nothing
-end
-
-@inline function _with_size(f, io::IOBuffer, sink, x, V)
-    if io.seekable
-        MAX_LENGTH_VARINT_BYTES = 5  # max size of a UInt32 as vbyte
-        initpos = position(io)
-        truncate(io, initpos + MAX_LENGTH_VARINT_BYTES) # 1.
-        seek(io, initpos + MAX_LENGTH_VARINT_BYTES)
-        f(sink, x, V) # e.g. _encode(io, x, Val{:zigzag}) # 2.
-        endpos = position(io)
-        data_len = endpos - initpos - MAX_LENGTH_VARINT_BYTES
-        seek(io, initpos)                  # 3.
-        vbyte_encode(io, UInt32(data_len)) # --||--
-        lenght_len = position(io) - initpos
-        unsafe_copyto!(io.data, initpos + lenght_len + 1, io.data, initpos + MAX_LENGTH_VARINT_BYTES + 1, data_len) # 4.
-        seek(io, initpos + lenght_len + data_len)
-        truncate(io, initpos + lenght_len + data_len)
+        ecoded_size = endpos - initpos - ecoded_size_len_guess
+        ecoded_size_len = _encoded_size(UInt32(ecoded_size))
+        # If our initial guess on encoded size of the size was wrong, then we need to move the encoded data
+        if ecoded_size_len_guess != ecoded_size_len
+            # If we guessed too small, we need to move the data to make room for the correct size
+            ecoded_size_len_guess < ecoded_size_len && truncate(io, initpos + ecoded_size_len + ecoded_size)
+            # Move the data right after the correct size
+            unsafe_copyto!(
+                io.data, initpos + ecoded_size_len + 1,
+                io.data, initpos + ecoded_size_len_guess + 1,
```

Comment:
I see this is the way it was before, but why do we need the `+1`s this time around? Above we seek to `initpos + encoded_size_len_guess` and start writing from there?

But I'm terrible with these off-by-one footguns so what you have is probably correct...

Reply:
This is because the `initpos` comes from `position(::IO)` which starts at zero (i.e. `position(IOBuffer()) == 0`).

>But I'm terrible with these off-by-one footguns so what you have is probably correct...

I don't want to turn this into a competition, but I think I'm way worse with off-by-one errors:)

Reply:
😁